### ISSUE TITLE
Update snowflake-sqlalchemy version dependency to ^1.4.3 and SQLAlchemy to ^1.4.42

### ISSUE THREAD

**Issue by tlento** [transform-data/metricflow#382]
We are currently stuck on SQLAlchemy versions between 1.4.27 and 1.4.41 because of a conflict with Snowflake's SQLAlchemy connector.

That conflict was resolved with Snowflake's release of 1.4.3, so we should update accordingly and push version restriction on SQLAlchemy to 1.4.42.

This has hard-to-reason-about downstream effects, so it's best left to a member of the core team.

**Comment by tlento** [comment]
Note - we're currently stuck on SqlAlchemy 1.4.27 due to the requirements associated with our bigquery-sqlalchemy dependency.

So this is blocked on our larger package resolution problems. #84 is likely the solution, as it could help us resolve both this issue and a known issue with dependency conflicts between the Databricks connector and the BigQuery SQLAlchemy package.

**Comment by callum-mcdata** [comment]
Hey folks! I just ran into this particular issue and was unable to connect to Snowflake at all given the deprecation of `_rfc_1738_quote`. Not entirely sure what was going on there so I decided to test it out on a clean install. 

I confirmed that neither `SQLAlchemy` or any of its snowflake derivatives were installed in my global installation. Here is the `pip list` outside of the venv:

<details>
<summary>List</summary>

```
agate                    1.6.3
attrs                    22.1.0
aws-error-utils          1.3.0
aws-sso-lib              1.13.0
Babel                    2.10.3
boto3                    1.26.52
botocore                 1.29.52
certifi                  2022.6.15
cffi                     1.15.1
charset-normalizer       2.1.1
click                    8.1.3
colorama                 0.4.5
dbt-extractor            0.4.1
distlib                  0.3.6
filelock                 3.8.0
future                   0.18.2
hologram                 0.0.15
idna                     3.3
iniconfig                1.1.1
isodate                  0.6.1
Jinja2                   3.1.2
jmespath                 1.0.1
jsonschema               3.2.0
leather                  0.3.4
Logbook                  1.5.3
MarkupSafe               2.1.1
mashumaro                3.0.4
minimal-snowplow-tracker 0.0.2
msgpack                  1.0.4
networkx                 2.8.6
packaging                21.3
parsedatetime            2.4
pathspec                 0.9.0
pip                      22.3.1
platformdirs             2.5.4
pluggy                   1.0.0
py                       1.11.0
pycparser                2.21
pyparsing                3.0.9
pyrsistent               0.18.1
pytest                   7.1.2
python-dateutil          2.8.2
python-slugify           6.1.2
pytimeparse              1.1.8
pytz                     2022.2.1
PyYAML                   6.0
requests                 2.28.1
s3transfer               0.6.0
setuptools               65.6.3
six                      1.16.0
sqlparse                 0.4.2
text-unidecode           1.3
tomli                    2.0.1
typing_extensions        4.3.0
urllib3                  1.26.12
virtualenv               20.16.7
Werkzeug                 2.2.2
wheel                    0.38.4
```
</details>

The thing that I noticed is that it appears the version of SQLAlchemy being installed is `1.4.46`, which causes the issue around `_rfc_1738_quote` not being present. This was blocking my from running `mf health-checks`.

<details>
<summary>List after pip install metricflow</summary>

```
asn1crypto                    1.5.1
attrs                         22.2.0
backoff                       2.2.1
cachetools                    5.3.0
certifi                       2022.12.7
cffi                          1.15.1
charset-normalizer            2.1.1
click                         8.1.3
colorama                      0.4.6
croniter                      1.3.8
cryptography                  38.0.4
databricks-sql-connector      2.0.3
duckdb                        0.3.4
duckdb-engine                 0.1.11
filelock                      3.9.0
future                        0.18.3
fuzzywuzzy                    0.18.0
gitdb                         4.0.10
GitPython                     3.1.30
google-api-core               2.11.0
google-auth                   2.16.0
google-cloud-bigquery         2.34.2
google-cloud-bigquery-storage 2.18.1
google-cloud-core             2.3.2
google-crc32c                 1.5.0
google-resumable-media        2.4.1
googleapis-common-protos      1.58.0
graphviz                      0.18.2
greenlet                      2.0.1
grpcio                        1.51.1
grpcio-status                 1.51.1
halo                          0.0.31
idna                          3.4
Jinja2                        3.1.2
jsonschema                    3.2.0
log-symbols                   0.0.14
MarkupSafe                    2.0.1
metricflow                    0.130.1
mo-dots                       4.22.21108
mo-future                     3.147.20327
mo-imports                    3.149.20327
mo-kwargs                     4.22.21108
mo-logs                       4.23.21108
monotonic                     1.6
more-itertools                8.10.0
moz-sql-parser                4.40.21126
numpy                         1.24.1
oscrypto                      1.3.0
packaging                     23.0
pandas                        1.5.3
pathspec                      0.11.0
pip                           22.3.1
proto-plus                    1.22.2
protobuf                      4.21.12
psycopg2                      2.9.5
pyarrow                       5.0.0
pyasn1                        0.4.8
pyasn1-modules                0.2.8
pycparser                     2.21
pycron                        3.0.0
pycryptodomex                 3.16.0
pydantic                      1.10.4
PyJWT                         2.6.0
pyOpenSSL                     22.1.0
pyrsistent                    0.19.3
python-dateutil               2.8.2
python-Levenshtein            0.12.2
pytz                          2022.7.1
PyYAML                        6.0
requests                      2.28.2
rsa                           4.9
ruamel.yaml                   0.17.21
ruamel.yaml.clib              0.2.7
rudder-sdk-python             1.0.6
setuptools                    65.6.3
six                           1.16.0
smmap                         5.0.0
snowflake-connector-python    2.9.0
snowflake-sqlalchemy          1.2.3
spinners                      0.0.24
SQLAlchemy                    1.4.46
sqlalchemy-bigquery           1.5.0
sqlalchemy-redshift           0.8.1
sqlalchemy2-stubs             0.0.2a32
tabulate                      0.8.9
termcolor                     2.2.0
thrift                        0.13.0
typing_extensions             4.4.0
update-checker                0.18.0
urllib3                       1.26.14
yamllint                      1.29.0
```

</details>

I was able to run `pip install snowflake-sqlalchemy==1.4.5` and resolve this issue but am unsure if it caused any downstream dependency issues - all I've tested so far is that health checks are passing.

**Comment by tlento** [comment]
> I was able to run pip install snowflake-sqlalchemy==1.4.5 and resolve this issue but am unsure if it caused any downstream dependency issues - all I've tested so far is that health checks are passing.

Ah, I think I see what happened here. MetricFlow has a version pin for `snowflake-sqlalchemy==1.2.3`, but no corresponding pin on the SQLAlchemy version. My guess is your `pip install snowflake-sqlalchemy==1.4.5` force-upgraded the package without regard for MetricFlow's dependency requirements, and happily it appears everything else still works.

Our next release, which is going out in a few minutes, has a temporary fix in place which pins the SQLAlchemy version to one compatible with the `snowflake-sqlalchemy 1.2.3` release. Now that I think about it, I don't know what happens if someone pip installs it over an existing later version of either of those packages, because I've never been able to understand how pip does dependency resolution. I'll try that out and see if I need to update our release notes with warnings.

This issue should resolve it all by updating our snowflake-sqlalchemy version to 1.4.x, which seems to be working for you already (so that's really nice, makes me think this issue won't be so hard to resolve after all). 

All this to say, I think we should prioritize this as the proper fix for this issue and bring our dependency versions more up to date.

**Comment by tlento** [comment]
It turns out pip's "downgrade package versions for compatibility" mechanisms are best effort, since it preserved my SQLAlchemy 1.4.46 installation instead of downgrading it to 1.4.42. Given this, I think we should hold the release until I can dig into this further, because the update we have in place might make people's existing SQLAlchemy/snowflake-sqlalchemy dependency situation materially worse unless they're always installing metricflow in a clean virtual environment.

Thanks for flagging this @callum-mcdata !

**Comment by tlento** [comment]
Fix is in main and will roll out with `0.140.0`
